### PR TITLE
*: reduce Key related memcpy

### PR DESCRIPTION
**Pull request by Connor1996**
## What have you changed? (mandatory)

- add a `take_encoded()` function of Key, so there is no need to call `encoded().to_vec()/to_owned()` anymore which cause memcpy.
- adjust `append_ts()` and `truncate_ts()` of Key to move instead of copying itself.
- remove `shrink_to_fit()` in decode function, because it maybe cause `rallocx`.

## What are the type of the changes? (mandatory)
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
run select count(*) manually

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

## Benchmark result if necessary (optional)
**before:**
mysql> select count(\*) from sbtest1;
+-----------+
| count(*)  |
+-----------+
| 100000000 |
+-----------+
1 row in set (12.12 sec)

**after:**
select count(\*) from sbtest1;
+-----------+
| count(*)  |
+-----------+
| 100000000 |
+-----------+
1 row in set (10.44 sec)

### CODE REVIEW

**Comment by breeswish** [comment]
How is the performance of these optimization individually? for example, seems that for the second optimization, it requires the caller to clone the key, which just balanced out the positive effect of the opt itself?

**Comment by Connor1996** [comment]
@breeswish not all callers need to clone, they just not involve in changes. Such like `Key::from_encoded(key.to_vec()).append_ts(0)`, it does benefit from optimization.

File: src/util/codec/bytes.rs [hunk 1]
```diff
@@ -23,7 +23,7 @@ const ENC_PADDING: [u8; ENC_GROUP_SIZE] = [0; ENC_GROUP_SIZE];
 
 // returns the maximum encoded bytes size.
 pub fn max_encoded_bytes_size(n: usize) -> usize {
-    (n / ENC_GROUP_SIZE + 1) * (ENC_GROUP_SIZE + 1)
+    (n / ENC_GROUP_SIZE + 1) * (ENC_GROUP_SIZE + 1) + number::U64_SIZE
```
Comment (siddontang) [inline src/util/codec/bytes.rs]: can you add a comment that why we need to add u64 size here?

File: src/storage/types.rs [hunk 2]
```diff
@@ -109,21 +124,17 @@ impl Key {
             // in the core storage engine layer.
             Err(codec::Error::KeyLength)
         } else {
-            let mut ts = &self.0[len - number::U64_SIZE..];
-            Ok(number::decode_u64_desc(&mut ts)?)
+            Ok(decode_ts(&self.0)?)
```
Comment (zhangjinpeng1987) [inline src/storage/types.rs]: So here judge the length twice?

File: src/storage/types.rs [hunk 3]
```diff
@@ -109,21 +124,17 @@ impl Key {
...
     /// Preconditions: the caller must ensure this is actually a timestamped key.
-    pub fn truncate_ts(&self) -> Result<Key, codec::Error> {
+    pub fn truncate_ts(mut self) -> Result<Key, codec::Error> {
         let len = self.0.len();
-        if len < number::U64_SIZE {
-            // TODO: (the same as above)
-            return Err(codec::Error::KeyLength);
-        }
-        Ok(Key::from_encoded(truncate_ts(&self.0).to_vec()))
+        self.0.truncate(len - number::U64_SIZE);
```
Comment (zhangjinpeng1987) [inline src/storage/types.rs]: Need check the length here.

**Comment by breeswish** [comment]
I think this PR is pretty tricky, which only solves a surface problem. However, it does improve some performance. LGTM. But I would like to see a deeper refinement over our `Key` in future.

File: src/util/codec/bytes.rs [hunk 4]
```diff
@@ -193,7 +193,7 @@ pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
         if padding.iter().any(|x| *x != pad_byte) {
```
Comment (siddontang) [inline src/util/codec/bytes.rs]: do we need to alloc extra u64 size for the decode_bytes?

Comment (Connor1996) [reply src/util/codec/bytes.rs]: no need, decode_bytes() creates raw key which will not append ts.

File: src/storage/types.rs [hunk 5]
```diff
@@ -84,9 +94,14 @@ impl Key {
         &self.0
     }
 
+    /// Gets and moves the encoded representation of this key.
+    pub fn take_encoded(self) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/storage/types.rs]: `into` is like a type conversion, `take` is similar to protobuf convention that `xxx()` get the reference of value and `take_xxx()` move the value.

File: src/util/codec/bytes.rs (hunk 4 above)
Comment (Connor1996) [reply src/util/codec/bytes.rs]: no need, seems that decode_bytes() will not reach and push the appended ts bytes

**Comment by siddontang** [comment]
@breeswish 

I think this PR only allocates extra capacity to avoid re-allocation for the timestamp. it is fine now. 

But later I suggest you and @Connor1996 refine the key implementation. 

LGTM now.

**Comment by breeswish** [comment]
@siddontang  @MyonKeminta is also interested in the key refinement. Also we would like to introduce a type of timestamped key so that `unwrap()` and runtime `assert!()` will not needed. Maybe we can discuss about what we can do later together.

**Comment by MyonKeminta** [comment]
@breeswish I'm afraid it might be a long story. So for now I think it's just ok to accept this PR.

**Comment by Connor1996** [comment]
@breeswish @MyonKeminta I have tried to introduce a timestamped key but it makes other places a little complex, since only WRITE_CF use timestamped key but others not which makes some functions type-mismatch. It would be great to discuss with you later. And now I am trying to replace `Vec<u8>` with `bytes::Bytes` to reduce memcpy further.

**Pull request merged**